akka-actor/src/main/scala/akka/util/WildcardTreeIndex.scala

```diff
@@ -0,0 +1,54 @@
+/**
+ * Copyright (C) 2009-2016 Lightbend Inc. <http://www.lightbend.com>
+ */
+
+package akka.util
+
+import scala.annotation.tailrec
+import scala.collection.immutable.HashMap
+
+private[akka] final case class WildcardIndex[T](wildcardTree: WildcardTree[T] = WildcardTree[T](), doubleWildcardTree: WildcardTree[T] = WildcardTree[T]()) {
+
+  def insert(elems: Array[String], d: T): WildcardIndex[T] = elems.lastOption match {
+    case Some("**") ⇒ copy(doubleWildcardTree = doubleWildcardTree.insert(elems.iterator, d))
+    case Some(_)    ⇒ copy(wildcardTree = wildcardTree.insert(elems.iterator, d))
+    case _          ⇒ this
+  }
+
+  def find(elems: Iterable[String]): Option[T] =
+    wildcardTree.findWithSingleWildcard(elems.iterator)
+      .orElse(doubleWildcardTree.findWithTerminalDoubleWildcard(elems.iterator))
```

Comment:
Make a special case for empty WildcardTree to avoid the `elems.iterator` and traversal thereof. I think it's worth doing that for both `wildcardTree` and `doubleWildcardTree`